src/neosophia/agents/base.py

```diff
@@ -31,230 +32,321 @@ class Agent:
+
+                n = 1
+                # Truncate prompt if it's too long - probably a better way
+                # to do this to keep relevant information
+                while num_tokens > self.model_info.max_tokens:
+                    prompt_str = prompt_str[n:]
+                    n += 1
+                    num_tokens = autils.count_tokens(
+                        prompt_str, self.model_info.name)
+
+            end_time = round(time.time() - time_start, 2)
+            total_cost = round(self.input_cost + self.output_cost, 2)
             print(answer)
-            print(80 * '-')
+            nct = f'| Number of LLM Calls: {self.llm_calls}\n'
+            nct += f'| Time: {end_time}\n'
+            nct += f'| Input Cost: {round(self.input_cost, 2)}\n'
+            nct += f'| Output Cost: {round(self.output_cost, 2)}\n'
+            nct += f'| Total Cost: {total_cost}'
+
+            print('\n')
+            print(nct)
+            print('\n')
 
-    def extract_params(self, parsed_data: Dict, function_resources: Dict):
+            self.save_log()
+
+    def extract_params(self, parsed_data: Dict):
         """ Extract parameters from LLM response """
 
-        func_key = 'Function'
+        func_key = 'Tool'
         param_prefix = 'Parameter_'
 
-        function = None
+        # Get the tool from available tools. If this function returns None,
+        # then it must have chosen a tool that isn't available
+        tool = None
         if func_key in parsed_data:
-            if parsed_data[func_key] in self.function_calls:
-                function = self.function_calls[parsed_data[func_key]]
+            if parsed_data[func_key] in self.tools:
+                tool = self.tools[parsed_data[func_key]]
 
-        # Create a dictionary of arguments to be passed to teh function
+        # Create a dictionary of arguments to be passed to the function
         args = {}
         for key, value in parsed_data.items():
             if key.startswith(param_prefix):
                 param_name = value[0]
                 param_value = value[1]
-
-                if param_value in function_resources:
-                    param_value = function_resources[param_value]
-
                 param_type = value[2].replace(' ', '')
+                param_vr = value[3]
 
-                # Stripping out quotes differently when it's a query
-                if param_type == 'str' and param_name != 'query':
+                # The parameter is a reference to a Variable in self.variables
+                if 'reference' in param_vr:
+                    # Strip out any quotes that might be at the beginning/end
+                    if param_value[0] == "'" or param_value[0] == '"':
+                        param_value = param_value[1:-1]
+                    param_value = self.variables[param_value].value
+
+                # Parameter is a string but not a SQL query
+                elif param_type == 'str' and param_name != 'query':
```

Comment:
I still think that this logic for removing quotes should only apply to the beginning and ending. eg if the both the first and last char are a single quote or both first and last are a double quote, remove first and last character. As written this will eliminate internal quotes, apostrophes, etc